torchmultimodal/modules/optimizers/anyprecision.py

```diff
@@ -0,0 +1,185 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+# AnyPrecisionAdamW: a flexible precision AdamW optimizer
+# with optional Kahan summation for high precision weight updates.
+# Allows direct control over momentum, variance and auxiliary compensation
+# buffer dtypes.
+# Optional Kahan summation is used to offset precision reduction for
+# the weight updates. This allows full training in BFloat16 (equal or
+# better than FP32 results in many cases) due to high precision weight upates.
+
+from typing import Any, Dict, Iterable, Tuple, Union
+
+import torch
+from torch.optim.optimizer import Optimizer
+from typing_extensions import TypeAlias
+
+params_t: TypeAlias = Union[Iterable[torch.Tensor], Iterable[Dict[str, Any]]]
+
+
+class AnyPrecisionAdamW(Optimizer):
+    def __init__(
+        self,
+        params: params_t,
+        lr: float = 1e-3,
+        betas: Tuple[float, float] = (0.9, 0.999),
+        eps: float = 1e-8,
+        weight_decay: float = 0.0,
+        use_kahan_summation: bool = False,
+        momentum_dtype: torch.dtype = torch.float32,
+        variance_dtype: torch.dtype = torch.bfloat16,
+        compensation_buffer_dtype: torch.dtype = torch.bfloat16,
+    ) -> None:
+        """
```

Comment:
good point, this will help ensure people don't think it's two different impls. 
I've added a note and pointer to the original TorchDistX so people know it's the same, just moved for easier access.